File: backend/src/main/java/org/example/backend/service/FriendshipService.java
package org.example.backend.service;

import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.example.backend.entity.Friendship;
import org.example.backend.entity.User;
import org.example.backend.exception.FriendshipException;
import org.example.backend.repository.FriendshipRepository;
import org.example.backend.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class FriendshipService {

    private final FriendshipRepository friendshipRepository;
    private final UserRepository userRepository;

    @Transactional
    public Friendship createFriendRequest(String requesterEmail, String receiverEmail) {
        // Validate users exist
        User requester = userRepository.findByEmail(requesterEmail)
                .orElseThrow(() -> new EntityNotFoundException("Người gửi không tồn tại"));
        User receiver = userRepository.findByEmail(receiverEmail)
                .orElseThrow(() -> new EntityNotFoundException("Người nhận không tồn tại"));

        // Check if friendship already exists
        Optional<Friendship> existingFriendship = friendshipRepository.findByRequesterAndReceiver(requester, receiver);
        if (existingFriendship.isPresent()) {
            Friendship friendship = existingFriendship.get();
            if (friendship.getStatus() == Friendship.FriendshipStatus.PENDING) {
                throw new FriendshipException("Đã gửi lời mời kết bạn trước đó");
            } else if (friendship.getStatus() == Friendship.FriendshipStatus.ACCEPTED && friendship.isActive()) {
                throw new FriendshipException("Đã là bạn bè");
            } else if (friendship.getStatus() == Friendship.FriendshipStatus.BLOCKED) {
                throw new FriendshipException("Không thể gửi lời mời kết bạn");
            } else if (friendship.getStatus() == Friendship.FriendshipStatus.REJECTED) {
                // Reuse existing friendship but reset to PENDING
                friendship.setStatus(Friendship.FriendshipStatus.PENDING);
                friendship.setActive(true);
                return friendshipRepository.save(friendship);
            }
        }

        // Check if reverse friendship exists and is blocked
        Optional<Friendship> reverseFriendship = friendshipRepository.findByRequesterAndReceiver(receiver, requester);
        if (reverseFriendship.isPresent() &&
                reverseFriendship.get().getStatus() == Friendship.FriendshipStatus.BLOCKED) {
            throw new FriendshipException("Không thể gửi lời mời kết bạn");
        }

        // If reverse friendship exists and is PENDING, accept it
        if (reverseFriendship.isPresent() &&
                reverseFriendship.get().getStatus() == Friendship.FriendshipStatus.PENDING) {
            Friendship friendship = reverseFriendship.get();
            friendship.setStatus(Friendship.FriendshipStatus.ACCEPTED);
            return friendshipRepository.save(friendship);
        }

        // Create new friendship request
        Friendship friendship = Friendship.builder()
                .requester(requester)
                .receiver(receiver)
                .status(Friendship.FriendshipStatus.PENDING)
                .active(true)
                .build();

        return friendshipRepository.save(friendship);
    }

    @Transactional
    public void cancelFriendRequest(String requesterEmail, String receiverEmail) {
        User requester = userRepository.findByEmail(requesterEmail)
                .orElseThrow(() -> new EntityNotFoundException("Người gửi không tồn tại"));
        User receiver = userRepository.findByEmail(receiverEmail)
                .orElseThrow(() -> new EntityNotFoundException("Người nhận không tồn tại"));

        Friendship friendship = friendshipRepository.findByRequesterAndReceiverAndStatus(
                requester, receiver, Friendship.FriendshipStatus.PENDING)
                .orElseThrow(() -> new FriendshipException("Không tìm thấy lời mời kết bạn"));

        friendshipRepository.delete(friendship);
    }

    @Transactional
    public Friendship acceptFriendRequest(Long friendshipId, String receiverEmail) {
        User receiver = userRepository.findByEmail(receiverEmail)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));

        Friendship friendship = friendshipRepository.findById(friendshipId)
                .orElseThrow(() -> new FriendshipException("Không tìm thấy yêu cầu kết bạn"));

        // Verify the current user is the receiver of this friend request
        if (!friendship.getReceiver().getId().equals(receiver.getId())) {
            throw new FriendshipException("Không có quyền chấp nhận yêu cầu kết bạn này");
        }

        if (friendship.getStatus() != Friendship.FriendshipStatus.PENDING) {
            throw new FriendshipException("Yêu cầu kết bạn không còn ở trạng thái chờ xác nhận");
        }

        friendship.setStatus(Friendship.FriendshipStatus.ACCEPTED);
        friendship.setAcceptedAt(LocalDateTime.now());
        return friendshipRepository.save(friendship);
    }

    @Transactional
    public Friendship rejectFriendRequest(Long friendshipId, String receiverEmail) {
        User receiver = userRepository.findByEmail(receiverEmail)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));

        Friendship friendship = friendshipRepository.findById(friendshipId)
                .orElseThrow(() -> new FriendshipException("Không tìm thấy yêu cầu kết bạn"));

        // Verify the current user is the receiver of this friend request
        if (!friendship.getReceiver().getId().equals(receiver.getId())) {
            throw new FriendshipException("Không có quyền từ chối yêu cầu kết bạn này");
        }

        if (friendship.getStatus() != Friendship.FriendshipStatus.PENDING) {
            throw new FriendshipException("Yêu cầu kết bạn không còn ở trạng thái chờ xác nhận");
        }

        friendship.setStatus(Friendship.FriendshipStatus.REJECTED);
        return friendshipRepository.save(friendship);
    }

    @Transactional
    public Friendship blockUser(String blockerEmail, Long blockeeId) {
        User blocker = userRepository.findByEmail(blockerEmail)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));
        User blockee = userRepository.findById(blockeeId)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng bị chặn không tồn tại"));

        // Check if there's an existing friendship
        Optional<Friendship> existingFriendship = friendshipRepository.findByRequesterAndReceiver(blocker, blockee);
        if (existingFriendship.isPresent()) {
            Friendship friendship = existingFriendship.get();
            friendship.setStatus(Friendship.FriendshipStatus.BLOCKED);
            return friendshipRepository.save(friendship);
        }

        // Check if there's a reverse friendship (convert it to blocker->blockee)
        Optional<Friendship> reverseFriendship = friendshipRepository.findByRequesterAndReceiver(blockee, blocker);
        if (reverseFriendship.isPresent()) {
            // Delete the reverse friendship
            friendshipRepository.delete(reverseFriendship.get());
        }

        // Create new block relationship
        Friendship friendship = Friendship.builder()
                .requester(blocker)
                .receiver(blockee)
                .status(Friendship.FriendshipStatus.BLOCKED)
                .active(true)
                .build();

        return friendshipRepository.save(friendship);
    }

    @Transactional
    public void unblockUser(String blockerEmail, Long blockeeId) {
        User blocker = userRepository.findByEmail(blockerEmail)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));
        User blockee = userRepository.findById(blockeeId)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng bị chặn không tồn tại"));

        Friendship friendship = friendshipRepository.findByRequesterAndReceiverAndStatus(
                blocker, blockee, Friendship.FriendshipStatus.BLOCKED)
                .orElseThrow(() -> new FriendshipException("Không tìm thấy người dùng bị chặn"));

        friendshipRepository.delete(friendship);
    }

    public List<Friendship> getPendingFriendRequests(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));

        return friendshipRepository.findByReceiverAndStatus(user, Friendship.FriendshipStatus.PENDING);
    }

    public List<Friendship> getSentFriendRequests(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));

        return friendshipRepository.findByRequesterAndStatus(user, Friendship.FriendshipStatus.PENDING);
    }

    public List<Friendship> getFriends(Long userId) {
        userRepository.findById(userId)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));

        // Tìm tất cả bạn bè từ cả hai hướng (người dùng là người gửi hoặc người nhận)
        List<Friendship> friends = friendshipRepository.findByRequesterIdOrReceiverId(userId, userId)
                .stream()
                .filter(friendship -> friendship.getStatus() == Friendship.FriendshipStatus.ACCEPTED &&
                        friendship.isActive())
                .collect(Collectors.toList());

        // Debug log
        System.out.println("User " + userId + " has " + friends.size() + " friends");
        friends.forEach(f -> System.out.println("Friend: " +
                (f.getRequester().getId().equals(userId) ? f.getReceiver().getId() : f.getRequester().getId())));

        return friends;
    }

    public List<Friendship> getBlockedUsers(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));

        return friendshipRepository.findByRequesterAndStatus(user, Friendship.FriendshipStatus.BLOCKED);
    }

    @Transactional
    public void removeFriend(Long friendshipId, String userEmail) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));

        Friendship friendship = friendshipRepository.findById(friendshipId)
                .orElseThrow(() -> new FriendshipException("Không tìm thấy mối quan hệ bạn bè"));

        // Verify the current user is part of this friendship
        if (!friendship.getRequester().getId().equals(user.getId()) &&
                !friendship.getReceiver().getId().equals(user.getId())) {
            throw new FriendshipException("Không có quyền xóa mối quan hệ bạn bè này");
        }

        if (friendship.getStatus() != Friendship.FriendshipStatus.ACCEPTED) {
            throw new FriendshipException("Không phải là bạn bè");
        }

        friendshipRepository.delete(friendship);
    }

    public List<Map<String, Object>> getFriendSuggestions(Long userId) {
        User currentUser = userRepository.findById(userId)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng không tồn tại"));

        // Get all users
        List<User> allUsers = userRepository.findAll();

        // Get current friends
        List<Friendship> friends = getFriends(userId);
        Set<Long> friendIds = new HashSet<>();

        for (Friendship friendship : friends) {
            if (friendship.getRequester().getId().equals(userId)) {
                friendIds.add(friendship.getReceiver().getId());
            } else {
                friendIds.add(friendship.getRequester().getId());
            }
        }

        // Get sent requests
        List<Friendship> sentRequests = getSentFriendRequests(userId);
        Set<Long> sentRequestIds = sentRequests.stream()
                .map(request -> request.getReceiver().getId())
                .collect(Collectors.toSet());

        // Get received requests
        List<Friendship> receivedRequests = getPendingFriendRequests(userId);
        Set<Long> receivedRequestIds = receivedRequests.stream()
                .map(request -> request.getRequester().getId())
                .collect(Collectors.toSet());

        // Get blocked users
        List<Friendship> blockedUsers = getBlockedUsers(userId);
        Set<Long> blockedUserIds = blockedUsers.stream()
                .map(blocked -> blocked.getReceiver().getId())
                .collect(Collectors.toSet());

        // Get users who blocked current user
        List<Friendship> blockedBy = friendshipRepository.findByReceiverAndStatus(currentUser,
                Friendship.FriendshipStatus.BLOCKED);
        Set<Long> blockedByIds = blockedBy.stream()
                .map(blocked -> blocked.getRequester().getId())
                .collect(Collectors.toSet());

        // Filter users for suggestions
        return allUsers.stream()
                .filter(user -> !user.getId().equals(userId)) // Exclude current user
                .filter(user -> !friendIds.contains(user.getId())) // Exclude friends
                .filter(user -> !sentRequestIds.contains(user.getId())) // Exclude sent requests
                .filter(user -> !receivedRequestIds.contains(user.getId())) // Exclude received requests
                .filter(user -> !blockedUserIds.contains(user.getId())) // Exclude blocked users
                .filter(user -> !blockedByIds.contains(user.getId())) // Exclude users who blocked current user
                .map(user -> {
                    Map<String, Object> userMap = new HashMap<>();
                    userMap.put("id", user.getId());
                    userMap.put("firstName", user.getFirstName());
                    userMap.put("lastName", user.getLastName());
                    userMap.put("email", user.getEmail());
                    userMap.put("avatar", user.getAvatar());

                    // Calculate mutual friends
                    List<Friendship> userFriends = getFriends(user.getId());
                    Set<Long> userFriendIds = new HashSet<>();

                    for (Friendship friendship : userFriends) {
                        if (friendship.getRequester().getId().equals(user.getId())) {
                            userFriendIds.add(friendship.getReceiver().getId());
                        } else {
                            userFriendIds.add(friendship.getRequester().getId());
                        }
                    }

                    long mutualFriends = userFriendIds.stream()
                            .filter(friendIds::contains)
                            .count();

                    userMap.put("mutualFriends", mutualFriends);
                    return userMap;
                })
                .collect(Collectors.toList());
    }

    public Map<String, Object> getFriendshipStatus(Long currentUserId, Long otherUserId) {
        User currentUser = userRepository.findById(currentUserId)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng hiện tại không tồn tại"));
        User otherUser = userRepository.findById(otherUserId)
                .orElseThrow(() -> new EntityNotFoundException("Người dùng khác không tồn tại"));

        Map<String, Object> result = new HashMap<>();
        result.put("status", "NOT_FRIENDS");
        result.put("friendshipId", null);

        // Check if current user sent a request to other user
        Optional<Friendship> sentRequest = friendshipRepository.findByRequesterAndReceiver(currentUser, otherUser);
        if (sentRequest.isPresent()) {
            result.put("status", sentRequest.get().getStatus().toString());
            result.put("friendshipId", sentRequest.get().getId());
            return result;
        }

        // Check if other user sent a request to current user
        Optional<Friendship> receivedRequest = friendshipRepository.findByRequesterAndReceiver(otherUser, currentUser);
        if (receivedRequest.isPresent()) {
            String status = receivedRequest.get().getStatus().toString();
            if (status.equals("PENDING")) {
                result.put("status", "PENDING_RECEIVED");
            } else {
                result.put("status", status);
            }
            result.put("friendshipId", receivedRequest.get().getId());
            return result;
        }

        return result;
    }

    public User getOtherUserInFriendship(Long friendshipId, Long userId) {
        Friendship friendship = friendshipRepository.findById(friendshipId)
                .orElseThrow(() -> new RuntimeException("Friendship not found"));

        if (friendship.getRequester().getId().equals(userId)) {
            return friendship.getReceiver();
        } else if (friendship.getReceiver().getId().equals(userId)) {
            return friendship.getRequester();
        } else {
            throw new RuntimeException("User is not part of this friendship");
        }
    }
}

File: frontend/src/services/friendshipService.ts
import axios from "axios";
import { api } from "../configs/api";
import { Friendship } from "../types/friendship";
import { User } from "../types/user";

interface FriendshipStatus {
  status: string;
  friendshipId: number | null;
}

class FriendshipService {
  // Gửi lời mời kết bạn
  async sendFriendRequest(email: string): Promise<Friendship> {
    try {
      const response = await api.post(`/friendships/request`, {
        receiverEmail: email,
      });
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 409) {
          throw new Error("Bạn đã gửi lời mời kết bạn cho người này rồi");
        }
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message || "Không thể gửi lời mời kết bạn"
        );
      }
      throw new Error("Đã xảy ra lỗi khi gửi lời mời kết bạn");
    }
  }

  // Hủy lời mời kết bạn đã gửi
  async cancelFriendRequest(email: string): Promise<void> {
    try {
      await api.delete(`/friendships/request`, {
        data: { receiverEmail: email },
      });
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message || "Không thể hủy lời mời kết bạn"
        );
      }
      throw new Error("Đã xảy ra lỗi khi hủy lời mời kết bạn");
    }
  }

  // Chấp nhận lời mời kết bạn
  async acceptFriendRequest(friendshipId: number): Promise<Friendship> {
    try {
      const response = await api.post(`/friendships/accept/${friendshipId}`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message || "Không thể chấp nhận lời mời kết bạn"
        );
      }
      throw new Error("Đã xảy ra lỗi khi chấp nhận lời mời kết bạn");
    }
  }

  // Từ chối lời mời kết bạn
  async rejectFriendRequest(friendshipId: number): Promise<Friendship> {
    try {
      const response = await api.post(`/friendships/reject/${friendshipId}`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message || "Không thể từ chối lời mời kết bạn"
        );
      }
      throw new Error("Đã xảy ra lỗi khi từ chối lời mời kết bạn");
    }
  }

  // Chặn người dùng
  async blockUser(userId: number): Promise<Friendship> {
    try {
      const response = await api.post(`/friendships/block/${userId}`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message || "Không thể chặn người dùng"
        );
      }
      throw new Error("Đã xảy ra lỗi khi chặn người dùng");
    }
  }

  // Bỏ chặn người dùng
  async unblockUser(userId: number): Promise<void> {
    try {
      await api.post(`/friendships/unblock/${userId}`);
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message || "Không thể bỏ chặn người dùng"
        );
      }
      throw new Error("Đã xảy ra lỗi khi bỏ chặn người dùng");
    }
  }

  // Lấy danh sách lời mời kết bạn đã nhận
  async getPendingFriendRequests(): Promise<Friendship[]> {
    try {
      const response = await api.get(`/friendships/pending`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message ||
            "Không thể lấy danh sách lời mời kết bạn"
        );
      }
      throw new Error("Đã xảy ra lỗi khi lấy danh sách lời mời kết bạn");
    }
  }

  // Lấy danh sách lời mời kết bạn đã gửi
  async getSentFriendRequests(): Promise<Friendship[]> {
    try {
      const response = await api.get(`/friendships/sent`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message ||
            "Không thể lấy danh sách lời mời đã gửi"
        );
      }
      throw new Error("Đã xảy ra lỗi khi lấy danh sách lời mời đã gửi");
    }
  }

  // Lấy danh sách bạn bè
  async getFriends(): Promise<Friendship[]> {
    try {
      const response = await api.get(`/friendships/friends`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message || "Không thể lấy danh sách bạn bè"
        );
      }
      throw new Error("Đã xảy ra lỗi khi lấy danh sách bạn bè");
    }
  }

  // Lấy danh sách người dùng đã chặn
  async getBlockedUsers(): Promise<Friendship[]> {
    try {
      const response = await api.get(`/friendships/blocked`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message ||
            "Không thể lấy danh sách người dùng đã chặn"
        );
      }
      throw new Error("Đã xảy ra lỗi khi lấy danh sách người dùng đã chặn");
    }
  }

  // Lấy danh sách gợi ý kết bạn
  async getFriendSuggestions(): Promise<User[]> {
    try {
      const response = await api.get(`/friendships/suggestions`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message ||
            "Không thể lấy danh sách gợi ý kết bạn"
        );
      }
      throw new Error("Đã xảy ra lỗi khi lấy danh sách gợi ý kết bạn");
    }
  }

  // Kiểm tra trạng thái kết bạn với một người dùng khác
  async getFriendshipStatus(userId: number): Promise<FriendshipStatus> {
    try {
      const response = await api.get(`/friendships/status/${userId}`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message ||
            "Không thể kiểm tra trạng thái kết bạn"
        );
      }
      throw new Error("Đã xảy ra lỗi khi kiểm tra trạng thái kết bạn");
    }
  }

  // Hủy kết bạn
  async removeFriend(friendshipId: number): Promise<void> {
    try {
      await api.delete(`/friendships/${friendshipId}`);
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        throw new Error(
          error.response?.data?.message || "Không thể hủy kết bạn"
        );
      }
      throw new Error("Đã xảy ra lỗi khi hủy kết bạn");
    }
  }

  // Kiểm tra xem có phải bạn bè hay không
  async areFriends(otherUserId: number): Promise<boolean> {
    try {
      const status = await this.getFriendshipStatus(otherUserId);
      return status.status === "ACCEPTED";
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 401) {
        localStorage.removeItem("accessToken");
        throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
      }
      return false;
    }
  }

  // Lấy danh sách bạn bè theo email
  async getFriendsByEmail(email: string): Promise<Friendship[]> {
    try {
      // Đầu tiên lấy userId từ email
      const userResponse = await api.get(
        `/users/email/${encodeURIComponent(email)}`
      );
      const userId = userResponse.data.id;
      console.log("User ID from email:", userId); // Debug log

      // Lấy danh sách bạn bè theo userId với status ACCEPTED
      const response = await api.get(
        `/friendships/user/${userId}/friends?status=ACCEPTED`
      );
      console.log("Friends API response:", response.data); // Debug log

      // Lọc và map lại dữ liệu để lấy thông tin người bạn
      const friends = response.data.map((friendship: Friendship) => {
        // Nếu user trong friendship là người đang xem, lấy người còn lại
        if (friendship.user.id === userId) {
          // Lấy ID của người bạn từ friendship
          const friendId = friendship.receiver.id;
          // Gọi API để lấy thông tin người bạn
          return api.get(`/users/${friendId}`).then((friendResponse) => ({
            ...friendship,
            user: friendResponse.data,
          }));
        }
        return Promise.resolve(friendship);
      });

      // Đợi tất cả các promise hoàn thành
      return Promise.all(friends);
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        if (error.response?.status === 404) {
          throw new Error("Không tìm thấy người dùng");
        }
        throw new Error(
          error.response?.data?.message || "Không thể lấy danh sách bạn bè"
        );
      }
      throw new Error("Đã xảy ra lỗi khi lấy danh sách bạn bè");
    }
  }

  // Lấy danh sách bạn bè theo userId
  async getFriendsByUserId(userId: number): Promise<Friendship[]> {
    try {
      // Thêm tham số status để chỉ lấy bạn bè đã chấp nhận
      const response = await api.get(
        `/friendships/user/${userId}/friends?status=ACCEPTED`
      );
      console.log("Friends API response:", response.data); // Debug log

      // Lọc và map lại dữ liệu để lấy thông tin người bạn
      const friends = response.data.map((friendship: Friendship) => {
        console.log("Processing friendship:", friendship); // Debug log

        // Kiểm tra xem friendship có requester và receiver không
        const friendId =
          friendship.requester?.id === userId
            ? friendship.receiver?.id
            : friendship.requester?.id;

        if (!friendId) {
          console.error(
            "Không tìm thấy ID người bạn trong friendship:",
            friendship
          );
          return Promise.resolve(friendship);
        }

        console.log("Getting friend info for ID:", friendId); // Debug log
        // Gọi API để lấy thông tin người bạn
        return api.get(`/users/${friendId}`).then((friendResponse) => {
          console.log("Friend info response:", friendResponse.data); // Debug log
          return {
            ...friendship,
            user: friendResponse.data,
          };
        });
      });

      // Đợi tất cả các promise hoàn thành
      return Promise.all(friends);
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          localStorage.removeItem("accessToken");
          throw new Error("Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại");
        }
        if (error.response?.status === 404) {
          throw new Error("Không tìm thấy người dùng");
        }
        throw new Error(
          error.response?.data?.message || "Không thể lấy danh sách bạn bè"
        );
      }
      throw new Error("Đã xảy ra lỗi khi lấy danh sách bạn bè");
    }
  }
}

export default new FriendshipService();
